README.md

```diff
@@ -0,0 +1,400 @@
+        let metadata = Unsize::target_metadata(core::ptr::metadata(ptr));
+        let untyped_data_ptr = (ptr as *const T).cast::<()>();
+        // SAFETY: [`Unsize`] demands that the return value of `Unsize::target_metadata` is valid to be used together with the data pointer to be re-interpreted as the unsized type and that `std::mem::size_of` on `U` will report the same size as the `T`.
+        unsafe { Arc::from_raw(core::ptr::from_raw_parts(untyped_data_ptr, metadata)) }
+    }
+}
+```
+
+Important to note is that `Unsize` impls are required to return metadata that make the unsized object report the same size as the source type.
+If that was not the case, the `Arc` impl above would be unsound, as its destructor would try to deallocate a smaller allocation than it initially owned.
+
+### A delegating `CoerceUnsized` impl
+
+Such an impl is used for wrapper like types, such as `Cell<T>` or `Pin<T>` where the impl is required to list a `CoerceUnsized` bound on the generic parameters of the wrapping type.
+An example impl for the `Cell<T>` type would be the following:
+```rs
+impl<T, U> CoerceUnsized<Cell<U>> for Cell<T>
+where
+    T: CoerceUnsized<U>
+{
+    fn coerce_unsized(self) -> Cell<U> {
+        Cell::new(self.into_inner().coerce_unsized())
+    }
+}
+```
+
+A delegating impl is not limited to struct types, as such another example implementation for `Option<T>` would be:
+```rs
+impl<T, U> CoerceUnsized<Option<U>> for Option<T>
+where
+    T: CoerceUnsized<U>,
+{
+    fn coerce_unsized(self) -> Option<U> {
+        match self {
+            Option::Some(t) => Option::Some(t.coerce_unsized()),
+            Option::None => Option::None,
+        }
+    }
+}
+```
+With such an impl, `Option<&[T; N]>` could coerce to `Option<&[T]>`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+## `Unsize`
+
+The new `Unsize` trait definition looks like the following:
+
+```rs
+/// # Safety
+///
+/// The implementation of [`Unsize::target_metadata`] must return metadata that
+/// - is valid for interpreting the `Self` type to `Target`, and
+/// - where using `core::mem::size_of` on the unsized object will report the same size as on the source object.
+pub unsafe trait Unsize<Target>
+where
+    Target: ?Sized,
+{
+    fn target_metadata(metadata: <Self as Pointee>::Metadata) -> <Target as Pointee>::Metadata;
+}
+```
+
+This trait allows specifying how to derive the metadata required for unsizing from the metadata of the source type `Self` or the compile time type information.
+
+## `CoerceUnsized`
+
+The new `CoerceUnsized` trait definition looks like the following:
+
+```rust
+pub trait CoerceUnsized<Target> {
+    fn coerce_unsized(self) -> Target;
+}
+```
+
+Implementations of this trait now specify how the coercion is done.
+This also drops the `?Sized` bound on `Target`, as returning unsized values is not possible currently and it being possible to relax the bound later on.
```

Comment:
```suggestion
This also drops the `?Sized` bound on `Target`, as returning unsized values is not possible currently. This can be relaxed without breakage in the future.
```